### PR TITLE
Check status of PreBackupPod Deployments after operator restart

### LINKED ISSUE #335: Backups are stuck when K8up is restarted during PreBackup
## Describe the bug

1. Have prebackups
2. Apply schedule
3. Start prebackup
4. Kill K8up before prebackups complete and backups begin
5. Backups get stuck, they're all in "waiting" condition for the prebackup

## Additional context

* Time between backups doesn't matter, because Deployments ensure that the replicaset/pod stays there forever. On the test cluster, some prebackup deployments were days old.
* This looks like another problem K8up has with internal state resp. long-running tasks in goroutines (see #137 )
* This problem could cascade into Out-of-memory kills of K8up because each Backup is being reconciled after 30s again (see #331 ) and thus completely fail all following backups.

## Logs

This is the logs **after** K8up restart. Notice the `starting pre backup pod` message, there is no more relevant logs after this
```console
2021-01-27T17:11:33.245Z	INFO	controller-runtime.manager.controller.restore	Starting workers	{"reconciler group": "backup.appuio.ch", "reconciler kind": "Restore", "worker count": 1}
2021-01-27T17:11:33.254Z	INFO	observer	new event on observed job	{"event": "running", "jobName": "default/schedule-test-prune-z9sn5"}
2021-01-27T17:11:33.258Z	INFO	observer	new event on observed job	{"event": "running", "jobName": "default/schedule-test-prune-b56dk"}
2021-01-27T17:11:33.262Z	INFO	observer	new event on observed job	{"event": "running", "jobName": "default/schedule-test-prune-fw4bd"}
2021-01-27T17:11:33.262Z	INFO	observer	new event on observed job	{"event": "failed", "jobName": "default/schedule-test-archive-hsg4p"}
2021-01-27T17:11:33.269Z	INFO	controllers.Schedule	Updated SyncConfig status.	{"schedule": "default/schedule-test"}
2021-01-27T17:11:33.939Z	INFO	controllers.Backup	starting pre backup pod	{"backup": "default/schedule-test-backup-dgcnr", "namespace": "default", "name": "mysqldump"}
2021-01-27T17:11:33.960Z	INFO	controllers.Backup	starting pre backup pod	{"backup": "default/schedule-test-backup-spjrl", "namespace": "default", "name": "mysqldump"}
2021-01-27T17:11:33.976Z	INFO	controllers.Backup	starting pre backup pod	{"backup": "default/schedule-test-backup-bwbfs", "namespace": "default", "name": "mysqldump"}
```

## Expected behavior
One of
* K8up should pick up where it left: Finish previous backups (fail or success) but don't fail next backups too.
* Alternatively, each Backup gets its own prebackup pod/job. But having 1 deployment for all backups is a risk that could make all following backups fail too.
* Instead of waiting asynchronously in the operator, the operator should start another pod/job with the same image that waits for a prebackup and then updates the Backup object so that the operator can continue with the next step. Goes further into the direction of "K8up as a CLI", but requires biggest change

## To Reproduce

1. Insert an init container to prebackup sample:
```yaml
$ cat config/samples/k8up_v1alpha1_prebackuppod.yaml
apiVersion: backup.appuio.ch/v1alpha1
kind: PreBackupPod
metadata:
  name: mysqldump
spec:
  backupCommand: mysqldump -u$USER -p$PW -h $DB_HOST --all-databases
  pod:
    spec:
      initContainers:
      - name: init
        image: mariadb
        command:
          - sleep
          - '20s'
      containers:
        - env:
            - name: USER
              value: dumper
            - name: PW
              value: topsecret
            - name: DB_HOST
              value: mariadb.example.com
          image: mariadb
          command:
            - 'sleep'
            - '9999999'
          imagePullPolicy: Always
          name: mysqldump
```
1. Preparation
```bash
make clean e2e-test # to start the K8up operator. kind-run doesn't work with prebackups
kubectl apply -f config/samples/k8up_v1alpha1_prebackuppod.yaml
kubectl apply -f config/samples/secrets.yaml
kubectl apply -f config/samples/k8up_v1alpha1_schedule.yaml
```
1. Restart
```bash
# watch for prebackup pods
kubectl -n default get pods -w
# as soon as there is the prebackup pod, kill the operator 
# (that's why we have added the 20s sleep in the init, to give you time to kill)
kubectl -n k8up-system delete pod -l app.kubernetes.io/name=e2e
```
1. verify stuck
```bash
kubectl get backups --sort-by metadata.creationTimestamp
NAME                         SCHEDULE REF    COMPLETION   PREBACKUP   AGE
...
schedule-test-backup-rpr75   schedule-test                Waiting     3m6s
schedule-test-backup-zqlkj   schedule-test                Waiting     2m6s
schedule-test-backup-4ktbm   schedule-test                Waiting     66s
schedule-test-backup-q5kdp   schedule-test                Waiting     6s

kubectl get pods       
NAME                              READY   STATUS             RESTARTS   AGE
mysqldump-5cd8f467b4-4925q        1/1     Running            0          19m
```

## Environment (please complete the following information):

- Image Version: master (rc5+)
- K8s Version: v1.20
- K8s Distribution: KIND

### PR DESCRIPTION
**Pull request by cimnine**
## Summary

Fixes #335 

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`,
      as they show up in the changelog
- [x] Update the documentation.
- [x] Update tests.
- [x] Link this PR to related issues.

### CODE REVIEW

**Comment by cimnine** [comment]
I've rebased this PR onto #333 since that other PR introduces quite some changes to the integration test infrastructure. I've also changed the base of this PR, so that it does not show the changes of #333 as changes of this PR.

**Comment by cimnine** [comment]
Replaced PreBackupPod's client creation logic with [seiso's implementation](https://github.com/appuio/seiso/blob/fddf46ce9081c232cb98084f22abe254fd889165/pkg/kubernetes/basic.go).
Rebased onto `master` and organised the commits.

File: prebackup/prebackup.go [hunk 1]
```diff
@@ -119,64 +122,98 @@ func (p *PreBackup) generateDeployments(templates *k8upv1alpha1.PreBackupPodList
...
-func (p *PreBackup) getClientset() (*kubernetes.Clientset, error) {
+func getKubeConfig() (*rest.Config, error) {
+	kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{},
+	)
+	restConfig, err := kubeconfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
```
Comment (ccremer) [inline prebackup/prebackup.go:208]: might as well return directly, right?
```suggestion
	return kubeconfig.ClientConfig()
```
if not, `return restconfig, err` should do it

File: prebackup/prebackup.go [hunk 2]
```diff
@@ -190,43 +227,76 @@ func (p *PreBackup) getClientset() (*kubernetes.Clientset, error) {
...
+			p.Log.Error(p.CTX.Err(), "unexpected end during deployment watch ", "deployment", fmt.Sprintf("%s/%s", p.Obj.GetMetaObject().GetNamespace(), p.Obj.GetMetaObject().GetName()))
+			return p.CTX.Err()
+		}
+	}
+}
+
+func (p *PreBackup) handleEvent(event watch.Event) (bool, error) {
+	deployment, ok := event.Object.(*appsv1.Deployment)
+	if !ok {
+		p.Log.V(2).Info("unexpected event during deployment watch ", "event source", event.Object, "event type", event.Type)
```
Comment (ccremer) [inline prebackup/prebackup.go:251]: Why log level 2?  level 1 is already "debug", is "debug" enough to display?
Please mind that currently K8up can only log level 0 or 1

File: prebackup/prebackup.go [hunk 3]
```diff
@@ -190,43 +227,76 @@ func (p *PreBackup) getClientset() (*kubernetes.Clientset, error) {
...
+
+	default:
+		p.Log.Info("unexpected event during deployment watch ", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name), "event type", event.Type)
 	}
+	return false, nil
+}
 
-	return nil
+func (p *PreBackup) isDeploymentReady(deployment *appsv1.Deployment) (bool, error) {
+	last := p.getLastDeploymentCondition(deployment)
```
Comment (ccremer) [inline prebackup/prebackup.go:289]: We might want to check whether the deployment has a deletion date but still existed at the time when doing GET.

Comment (ccremer) [reply prebackup/prebackup.go:284]: Let's hope for now that this is not a use case we encounter often.

**Comment by ccremer** [comment]
I can confirm with this PR I cannot reproduce the issue locally anymore :tada:

**Pull request merged**